### PR TITLE
Export meshes as OBJ

### PR DESCRIPTION
**Pull request by andburn**
Implemented basic OBJ exporting, works for the unity 4 & 5 bundles I have.

Elements not implemented:
- _topology_ triangles
- _compressed meshes_
- _half-floats_ (16 bits)

Don't know how widespread these missing features are, but I didn't encounter them.

### CODE REVIEW

File: unitypack/export.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+				out += "/"
+				if vn:
+					out += str(i)
+			out += " "
+		return out + "\n"
+
+	def export(self):
+		vns = self.vertices["normals"]
+		vts = self.vertices["uv1"]
+		out_str = ""
```
Comment (jleclanche) [inline unitypack/export.py]: This is very slow. Create a list `ret = []`, and do `return "\n".join(ret)`.

File: unitypack/export.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+		for idx in inds:
+			i = idx + 1 # OBJ indices 1 based
+			out += str(i)
+			if vt or vn:
+				out += "/"
+				if vt:
+					out += str(i)
+				out += "/"
+				if vn:
+					out += str(i)
```
Comment (jleclanche) [inline unitypack/export.py]: Im not sure this works correctly if only one of vt or vn is non-null.

File: unitypack/export.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+	def vec2_str(self, prefix, vec):
+		return "%s %s %s\n" % (prefix, vec.x, 1 - vec.y)
+
+	def vec3_str(self, prefix, vec):
+		return "%s %s %s %s\n" % (prefix, -vec.x, vec.y, vec.z)
+
+	def face_str(self, indices, vts, vns):
+		out = "f "
+		# reverse indices
+		inds = indices[::-1]
```
Comment (jleclanche) [inline unitypack/export.py]: You don't need a temporary variable, just do `for i in indices[::-1]`

File: unitypack/export.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+		vt = vts > 0
+		vn = vns > 0
+		for idx in inds:
+			i = idx + 1 # OBJ indices 1 based
```
Comment (jleclanche) [inline unitypack/export.py]: Same here, just do`i + 1`

File: unitypack/export.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+		"uv3" : uv3,
+		"uv4" : uv4,
+		"tangents" : tangents
+	}
+
+
+class OBJMesh:
+	def __init__(self, mesh):
+		self.mesh = mesh
+		if (self.mesh.mesh_compression):
```
Comment (jleclanche) [inline unitypack/export.py]: extraneous parentheses

File: unitypack/export.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+		self.g = self.y
+		self.b = self.z
+		self.a = self.w
+
+
+def get_num_streams(channels):
+	streams = []
+	# scan the channel's stream value for distinct entries
+	for c in channels:
+		if c['stream'] not in streams:
```
Comment (jleclanche) [inline unitypack/export.py]: Bad quoting ;)

File: unitypack/export.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+	def read_color(self, buf):
+		self.x = buf.read_ubyte()
+		self.y = buf.read_ubyte()
+		self.z = buf.read_ubyte()
+		self.w = buf.read_ubyte()
+		self.__alt_rgba()
+		return self
+
+	def __alt_rgba(self):
```
Comment (jleclanche) [inline unitypack/export.py]: what's this for?

**Comment by andburn** [comment]
Updated:
- change string concatenation to list joins
- remove single quotes
- remove `Vec4` rgba aliases (not used)
- remove unnecessary temporary variables
- fix possible trailing slash on face definitions, if normals were empty

File: unitypack/export.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+	streams = []
+	# scan the channel's stream value for distinct entries
+	for c in channels:
+		if c["stream"] not in streams:
+			streams.append(c["stream"])
+
+	return len(streams)
+
+
+def extract_indicies(mesh):
```
Comment (jleclanche) [inline unitypack/export.py]: `indices`

File: unitypack/export.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+			sub_indices.append(buf.read_uint16())
+		if sub.topology == 0:
+			sub_triangles.extend(sub_indices)
+		else:
+			raise NotImplementedError("(%s) topologies are not supported" % (mesh.name))
+
+		indices.append(sub_indices)
+		triangles.append(sub_triangles)
+
+	return {
```
Comment (jleclanche) [inline unitypack/export.py]: I don't think you should be creating a dict for the return here. you can just `return indices, triangles`

File: unitypack/export.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+	}
+
+
+class OBJMesh:
+	def __init__(self, mesh):
+		self.mesh = mesh
+		if self.mesh.mesh_compression:
+			# TODO handle compressed meshes
+			raise NotImplementedError("(%s) compressed meshes are not supported" % (mesh.name))
+		else:
```
Comment (jleclanche) [inline unitypack/export.py]: extraneous else clause

File: unitypack/export.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+
+		ret.extend(["\ng ", self.mesh.name, "\ns 1\n" ])
+
+		sub_count = len(self.mesh.submeshes)
+		sub_tri_all = self.indicies["triangles"]
+		verts_per_face = 3
+
+		for i in range(0, sub_count):
+			if sub_count == 1:
+				ret.extend(["usemtl ", self.mesh.name, "\n"])
```
Comment (jleclanche) [inline unitypack/export.py]: You should be doing `ret.append("usemtl %s\n" % (self.mesh.name))`

File: unitypack/export.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+			else:
+				ret.extend(["usemtl ", self.mesh.name, "_", str(i), "\n"])
```
Comment (jleclanche) [inline unitypack/export.py]: Same here.

File: unitypack/export.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+			self.indicies = extract_indicies(mesh)
+			self.vertices = extract_vertices(mesh)
+
+	def vec2_str(self, prefix, vec):
+		return "%s %s %s\n" % (prefix, vec.x, 1 - vec.y)
+
+	def vec3_str(self, prefix, vec):
+		return "%s %s %s %s\n" % (prefix, -vec.x, vec.y, vec.z)
+
+	def face_str(self, indices, vts_len, vns_len):
```
Comment (jleclanche) [inline unitypack/export.py]: All the `_str` functions can be staticmethods.

Comment (jleclanche) [inline unitypack/export.py]: Come to think of it, you can just implement `vecX_str` on the class itself, which will let you do `str(vec)`. Much simpler.

File: unitypack/export.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+			vts = self.vertices["uv2"]
+
+		for v in self.vertices["vertices"]:
+			ret.append(self.vec3_str("v", v))
+		for v in vns:
+			ret.append(self.vec3_str("vn", v))
+		for v in vts:
+			ret.append(self.vec2_str("vt", v))
+
+		ret.extend(["\ng ", self.mesh.name, "\ns 1\n" ])
```
Comment (jleclanche) [inline unitypack/export.py]: This is not very readable.

File: unitypack/export.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+						else:
+							tangents.append(Vec4().read(buf))
+					elif j == 6: # for unity 5+
+						uv4.append(Vec2().read(buf))
+					elif j == 7: # for unity 5+
+						tangents.append(Vec4().read(buf))
+		# TODO investigate possible alignment here, after each stream
+
+	return {
+		"vertices": vertices,
```
Comment (andburn) [inline unitypack/export.py]: Are you ok with using the dictionary here, or is there a more pythonic way of doing it?

Comment (jleclanche) [inline unitypack/export.py]: Well, pythonically there's two options:
- Create a struct-like class with those attributes (or even just a namedtuple)
- Return multiple values... but with the amount of values being returned here, it gets unreadable.

**Comment by andburn** [comment]
Updated:
- Comments addressed.
- Make `vecX_str` methods static (I think they fit better inside the `OBJMesh` class, as they are specific to the OBJ format).
- Added `MeshData` class to hold vertex and index info, moved top-level methods into it too.

**Comment by jleclanche** [comment]
> I think they fit better inside the OBJMesh class, as they are specific to the OBJ format

I wouldn't disagree if these were generic vectors but... just call them OBJVector2, OBJVector3 etc and you should be fine.

**Comment by andburn** [comment]
Ok.
One other thing. Its outputting the values to 15 decimal places, does rounding it down to 6-8 seem sensible?

**Comment by jleclanche** [comment]
We can figure that out another time.

**Comment by andburn** [comment]
Updated to use `__str__`

**Comment by jleclanche** [comment]
Thank you!

**Pull request merged**